sys/parsoid.js

```diff
@@ -125,6 +122,51 @@ class ParsoidService {
         };
     }
 
+    _initOpts(opts) {
+        this.options = opts = opts || {};
+        this.parsoidHost = opts.parsoidHost;
+        this.options.stash_ratelimit = opts.stash_ratelimit || 5;
+        this.options.grace_ttl = opts.grace_ttl || 86400;
+        this._blacklist = compileReRenderBlacklist(opts.rerenderBlacklist);
+        if (!opts.parsoidHost) {
+            throw new Error('Parsoid module: the option parsoidHost must be provided!');
+        }
+    }
+
+    _checkStashRate(hyper, req) {
+        if (!hyper.ratelimiter) {
+            return;
+        }
+        if (hyper._rootReq.headers['x-request-class'] !== 'external') {
+            return;
+        }
+        if (!((req.query && req.query.stash) || (req.body && req.body.stash))) {
+            return;
+        }
+        if (hyper._rootReq.headers['x-stash-rate-checked']) {
```

Comment:
What's the point of this? I don't think it would be called recursively and if it does, it's an error.

Reply:
There are cases where parsoid calls itself via the URI, such as https://github.com/wikimedia/restbase/blob/de620a7f897a1f1e0840b009bb34293d11984b31/sys/parsoid.js#L577 (there are others too), so this ensures the rate is checked only once.

Reply:
Yes and no. Actually that particular line you have shown is not calling itself, it's actually constructing an artificial request for code sharing. I wanna refactor it out. The only other case is calling the `pagebundle` endpoint, which I wanna get rid of too. So this is not needed.

Reply:
Ah you are correct. Ok, removing.